fedora_messaging/schema_utils.py

```diff
@@ -0,0 +1,90 @@
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+"""
+Utilities for message schemas.
+"""
+
+import collections
+from hashlib import sha256, md5
+
+from six.moves.urllib import parse
+
+try:
+    import libravatar
+except ImportError:
+    # We can do without this dependency.
+    libravatar = None
+
+
+# https://github.com/fedora-infra/fedmsg_meta_fedora_infrastructure/issues/320
+HARDCODED_AVATARS = {
+    "bodhi": "https://apps.fedoraproject.org/img/icons/bodhi-{size}.png",
+    "koschei": "https://apps.fedoraproject.org/img/icons/koschei-{size}.png",
+    # Taskotron may have a new logo at some point.  Check this out:
+    # https://mashaleonova.wordpress.com/2015/08/18/a-logo-for-taskotron/
+    # Ask tflink before actually putting this in place though.  we need
+    # a nice small square version.  It'll look great!
+    # In the meantime, we can use this temporary logo.
+    "taskotron": "https://apps.fedoraproject.org/img/icons/taskotron-{size}.png",
+}
+
+
+def avatar_url(username, size=64, default="retro"):
+    """Get the avatar URL of the provided username.
+
+    Args:
+        username (str): The username to get the avatar of.
+        size (int): Size of the avatar in pixels (it's a square).
+        default (str): Default avatar to return if not found.
+    Returns:
+        str: The URL to the avatar image.
+    """
+    try:
+        return HARDCODED_AVATARS[username].format(size=size)
+    except KeyError:
+        openid = "http://{}.id.fedoraproject.org/".format(username)
+        return libravatar_url(openid=openid, size=size, default=default)
+
+
+def libravatar_url(email=None, openid=None, size=64, default="retro", dns=False):
+    """Get the user's avatar from libravatar.
+
+    Either the user's email or openid must be provided.
+
+    Args:
+        email (str): the user's email
+        openid (str): the user's OpenID
+        size (int): Size of the avatar in pixels (it's a square).
+        default (str): Default avatar to return if not found.
+        dns (bool): Send a DNS SRV query to use Libravatar federation.
+    Returns:
+        str: The URL to the avatar image.
+    """
+    if dns and libravatar is not None:
+        # This makes an extra DNS SRV query, which can slow down our
+        # webapps. It is necessary for libravatar federation, though.
+        return libravatar.libravatar_url(
+            email=email, openid=openid, size=size, default=default
+        )
+    params = collections.OrderedDict([("s", size), ("d", default)])
+    query = parse.urlencode(params)
+    if email:
+        idhash = md5(email.encode("utf-8")).hexdigest()
+    elif openid:
+        idhash = sha256(openid.encode("utf-8")).hexdigest()
+    else:
+        raise TypeError("You must provide either the email or the openid.")
```

Comment:
This isn't documented in the docblock. Also, it's a minor thing, but I think ValueError is more appropriate here since ``None`` is a valid type for both email and openid, but they can't both be None at the same time.